### ISSUE TITLE
graceful shutdown not working

### ISSUE THREAD

**Issue by pekeler** [Unitech/pm2#1763]
I can't seem to get a graceful shutdown working with pm2. Bug, or am I doing something wrong?

``` javascript
'use strict';

let stopped = false;

function work() {
  console.log('working');
  !stopped && setTimeout(work, 1000);
}

function stop() {
  console.log('shutting down');
  stopped = true;
}

process.once('SIGINT', stop);   // CTRL-C
process.once('SIGTERM', stop);  // pm2 stop
process.on('beforeExit', () => console.log('exited cleanly :)'));

work();
```

Running this little app `node test.js` followed after 3 seconds by a CTRL-C produces:

```
working
working
working
^Cshutting down
working
exited cleanly :)
```

However, with pm2 it never exits and gets killed after PM2_KILL_TIMEOUT (5s in my case). So `pm2 start test.js` followed 3 second later by `pm2 stop test.js` produces:

```
PM2 Starting execution sequence in -fork mode- for app name:test id:1
PM2 App name:test id:1 online
test-1 working
test-1 working
test-1 working
PM2 Stopping app:test id:1
test-1 shutting down
PM2 Process with pid 6079 still not killed, retrying...
PM2 Process with pid 6079 still not killed, retrying...
...
PM2 Process with pid 6079 still not killed, retrying...
PM2 Process with pid 6079 still not killed, retrying...
test-1 working
PM2 Process with pid 6079 still not killed, retrying...
PM2 Process with pid 6079 still not killed, retrying...
...
PM2 Process with pid 6079 still not killed, retrying...
PM2 Process with pid 6079 still not killed, retrying...
PM2 Process with pid 6079 still alive after 5000ms, sending it SIGKILL now...
PM2 App name:test id:1 exited with code SIGKILL
PM2 Process with pid 6079 killed
```

**Comment by pekeler** [comment]
OK, so pm2 forks my app which means my app is responsible for calling process.exit upon completion of the graceful shutdown. Unfortunately, determining when the graceful shutdown is done is very much a non-trivial problem in my real app.

Is there any config to spawn instead of fork my app? (I don't need clustering, I need exactly one instance.)

**Comment by Unitech** [comment]
Ok found the issue, it's going to be published soon

**Comment by Unitech** [comment]
There is javascript loop interval latency monitoring when starting an application with PM2, this interval was not stopped when process was being killed

**Comment by pekeler** [comment]
Thanks!

**Comment by Unitech** [comment]
https://github.com/keymetrics/pmx/commit/e60b051f7357ca46a34f60493ba5c64d785708d9

**Comment by aaronroberson** [comment]
@Unitech That link is to pmx and not pm2. Also, npm cannot find 0.15.11 it only finds 0.15.10. Has this update been published?

**Comment by ggarret** [comment]
An answer for @aaronroberson ? I've a same issue.

**Comment by Unitech** [comment]
SIGINT handle now works as expected (it will be called every time the process get exited).

To try it:

```
$ npm install pm2@next -g
$ pm2 update
```

**Comment by Unitech** [comment]
Patch available on pm2@2.1.4 (main):

```bash
$ npm install pm2 -g
$ pm2 update
```
